File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        
        char sign = '+';
        int index = 0;
        int length = s.length();
        long result = 0;

        while(index < length && s.charAt(index) == ' ')
            index++;
        
        if(index < length && s.charAt(index) == '+')
            index++;
        else if(index < length && s.charAt(index) == '-')
        {
            sign = '-';
            index++;
        }

        while(index < length)
        {
            if(!Character.isDigit(s.charAt(index)))
                break;
            else if(sign == '+' && result > Integer.MAX_VALUE)
            {
                result = Integer.MAX_VALUE;
                break;
            }
            else if(sign == '-' && -result < Integer.MIN_VALUE)
            {
                result = Integer.MIN_VALUE;
                break;
            }
            else
            {
                result = result * 10 + (s.charAt(index) - '0');
            }

            index++;
        }
        
        if(sign == '+' && result > Integer.MAX_VALUE)
            result = Integer.MAX_VALUE;
        else if(sign == '-' && -result < Integer.MIN_VALUE)
            result = Integer.MIN_VALUE;

        return sign == '-' ? -(int)(result) : (int)(result);
    }
}

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.js
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function(s) {
    
    let length = s.length;
    let index = 0;
    let result = 0; 
    let sign = '+';
    const MAX_VALUE = Math.pow(2, 31) - 1;
    const MIN_VALUE  = -Math.pow(2,31);

    while(index < length && s[index] == ' ')
        index++;
    
    if(s[index] === '+' || s[index] === '-')
    {
        sign = s[index];
        index++;
    }

    while(index < length)
    {
        if(isNaN(+s[index])|| s[index] === ' ')
            break;
        else if(sign === '+' && result > MAX_VALUE)
            return MAX_VALUE;
        else if(sign === '-' && -result < MIN_VALUE)
            return MIN_VALUE;
        else
            result = result * 10 + Number(s.charAt(index));

        index++;
    }

    if(sign === '+' && result > MAX_VALUE)
        return MAX_VALUE;
    else if(sign === '-' && -result < MIN_VALUE)
        return MIN_VALUE;

    return sign === '-' ? -result : result;
};